### DISCUSSION TITLE
Crashing in production

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#3166] [Q&A]
Is it possible that `dao` can be nil? It's the third time crashing the production server

```go
func (dao *Dao) lockRetry(op func(retryDao *Dao) error) error {
	retryDao := dao

	return baseLockRetry(func(attempt int) error {
		if attempt == 2 {
			// assign new Dao without the before hooks to avoid triggering
			// the already fired before callbacks multiple times
			retryDao = NewMultiDB(dao.concurrentDB, dao.nonconcurrentDB)
			retryDao.AfterCreateFunc = dao.AfterCreateFunc
			retryDao.AfterUpdateFunc = dao.AfterUpdateFunc
			retryDao.AfterDeleteFunc = dao.AfterDeleteFunc
		}

		return op(retryDao)
	}, dao.MaxLockRetries)
}
```

```logtalk
2023/08/20 10:29:17 http: Accept error: accept tcp 127.0.0.1:80: accept4: too many open files; retrying in 1s
2023/08/20 10:29:18 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
2023/08/20 10:29:18 goroutine 133274210 [running]:
runtime/debug.Stack()
	/Users/oggy/sdk/go1.20.7/src/runtime/debug/stack.go:24 +0x65
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/tools/routine/routine.go:26 +0x5d
panic({0x15b6c60, 0x2e6e330})
	/Users/oggy/sdk/go1.20.7/src/runtime/panic.go:884 +0x213
github.com/pocketbase/pocketbase/daos.(*Dao).lockRetry(0x0?, 0x0?)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/daos/base.go:338 +0x45
github.com/pocketbase/pocketbase/daos.(*Dao).Save(0xc007821ba0?, {0x2413ca8, 0xc02ba6b520})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/daos/base.go:223 +0x72
github.com/pocketbase/pocketbase/daos.(*Dao).SaveRequest(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/daos/request.go:69
github.com/pocketbase/pocketbase/apis.ActivityLogger.func1.1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/apis/middlewares.go:341 +0x58
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/tools/routine/routine.go:30 +0x8c
created by github.com/pocketbase/pocketbase/tools/routine.FireAndForget
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/tools/routine/routine.go:18 +0xac
```

**Comment by ganigeorgiev** [comment]
This is the same as in your other post in https://github.com/pocketbase/pocketbase/discussions/3067 (_and probably https://github.com/pocketbase/pocketbase/discussions/3093_).

It can't be nil unless the db connection wasn't initialized for some reason (or was abruptly closed).

In this specific case it seems that you've reached the file descriptors limit.
If you have a systemd service double check the `LimitNOFILE` configuration.
You can also check the limits for your OS user with `ulimit -a`.

It is also possible the error to be caused by something totally else but I'm not sure how to reproduce it. The last time I checked I couldn't reproduce it locally nor I wasn't able to find any panic logs on the live demo instance so I suspect that it is caused by something in your deployment env or custom code.

If you manage to reproduce it consistently, feel free to provide reproducible steps I'll try to investigate it in more details.

_Side-note: I also recommend to update to the latest PocketBase version since bugfixes are not applied for older releases._

**Reply by Oggy4th** [reply]
@ganigeorgiev

I checked dao in lockRetry function. It gives nil pointer error when dao is nil. I don't know why and how to fix it? 

```go
func (dao *Dao) lockRetry(op func(retryDao *Dao) error) error {
	if dao == nil {
		fmt.Println("DATE: ", time.Now().UTC().Format("2006-01-02 15:04:05"), ": lockRetry: dao is nil")
		return errors.New("Dao is nil")
	}

	retryDao := dao

	return baseLockRetry(func(attempt int) error {
		if attempt == 2 {
			// assign new Dao without the before hooks to avoid triggering
			// the already fired before callbacks multiple times
			retryDao = NewMultiDB(dao.concurrentDB, dao.nonconcurrentDB)
			fmt.Println("DATE: ", time.Now().UTC().Format("2006-01-02 15:04:05"), "lockRetry: dao.AfterCreateFunc ", dao.AfterCreateFunc)
			retryDao.AfterCreateFunc = dao.AfterCreateFunc
			retryDao.AfterUpdateFunc = dao.AfterUpdateFunc
			retryDao.AfterDeleteFunc = dao.AfterDeleteFunc
		}

		return op(retryDao)
	}, dao.MaxLockRetries)
}
```

```logtalk
2023/08/27 15:30:01 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
DATE:  2023-08-27 15:30:01 : lockRetry: dao is nil
2023/08/27 15:30:01 goroutine 129528300 [running]:
runtime/debug.Stack()
	/Users/oggy/sdk/go1.21.0/src/runtime/debug/stack.go:24 +0x5e
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/tools/routine/routine.go:27 +0x172
panic({0x156e640?, 0x2ee10e0?})
	/Users/oggy/sdk/go1.21.0/src/runtime/panic.go:914 +0x21f
main.main.func14.ActivityLogger.func7.1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/apis/middlewares.go:351 +0x247
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/tools/routine/routine.go:31 +0x87
created by github.com/pocketbase/pocketbase/tools/routine.FireAndForget in goroutine 129475241
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.16.9/tools/routine/routine.go:18 +0xa5
```

**Reply by ganigeorgiev** [reply]
See the comment from above:

> It can't be nil unless the db connection wasn't initialized for some reason (or was abruptly closed).

It could be also something completely different, I don't know.

Please try to create **minimal** reproducible repo with your Go hooks and I'll have a look, but as of now I can only make guesses.

**Reply by ganigeorgiev** [reply]
I've tried to reproduce it locally but without success.

The only scenario I can think of where the `app.Dao()` could end up `nil` is if the application is in the process of being terminated which triggers the `ResetBootstrapState()` while in the same time a retry operation is attempting to run. We can litter the code to stop the error "bleeding" but in this case it doesn't matter really and it could be safely ignored (_and once again notice that this is a "handled" panic, hence the "RECOVERED FROM PANIC" message_).

Again double check the ulimit/`LimitNOFILE` option if the connections limit is the cause of the application restart(s).

This is as far as I can help without more information about your setup, whether there is another application that is draining the connections/file descriptors or what custom Go code you run.

**Reply by ganigeorgiev** [reply]
To further clarify - to confirm that app termination is the cause you can check your logs and inspect whether there is a "Server started at ..." message shortly after the above error (in case auto restarts are enabled).

But again - **the error is NOT the cause of your app "crashing" because the panic is "handled"**, we just print the error message. The trigger for the error message originates from somewhere else.
The app could be terminated by the OS for various reasons - OOM, change of files permissions or as in your initial reported error - "too many open files" error.

**Reply by Oggy4th** [reply]
My speculation is that LogsDao using Nonconcurrent DB to save the log but failing or getting stuck in a loop with another hook (ping-pong) and locking the DB, the app is not really crashing it's just getting locked out, so no new requests get handled. It is clear that even after the "crash" cpu utilization rising up to around 90%

**Reply by ganigeorgiev** [reply]
The above assumption doesn't really make sense to me and I don't think LogsDao is related to the "stuck in a loop" as you refer it. The nonconcurrent db just ensures that only 1 db operation can occur at a time and it is used only for writes and transactions (_this is also the builtin SQLite WAL behavior; in our case it is implemented on application level as an extra measure to minimize the SQLITE_BUSY errors and to separate the read operations in order to keep the used file descriptors number low_).

The "stuck in a loop" comes from the fact that the regular application flow is interrupted by the OS - see the "too many open files;" error.

Again, the only way I'm able to reproduce partially the above panic is if the application is in the process of being terminated. This is the only thing that can trigger the `ResetBootstrapState()` which will set the `Dao` to `nil`.

Please read the described scenario in https://github.com/pocketbase/pocketbase/discussions/3166#discussioncomment-6852550. If that doesn't match with your use case, please provide more details or at least some reproducible steps, otherwise I really don't know how to help you.

I've asked several times in this discussion, in https://github.com/pocketbase/pocketbase/discussions/3067 and https://github.com/pocketbase/pocketbase/discussions/3093 but you refuse to provide more details or reproducible steps, so I'm going to close the discussion since we are repeating the same thing over and over.

**Reply by Oggy4th** [reply]
I have no idea why it occurs, seems so random to me sometimes it doesn't crash for a week, other times it crashes after 30mins of running, that is why I can't create minimal reproducible code. I'll investigate it further if I find something specific, I'll let you know.
